experiments/2018-05-23-rust-on-android-JNA.md

```diff
@@ -0,0 +1,198 @@
+---
+title: Building and Deploying a Rust library on Android using JNA
+layout: text
+---
+
+# Building and Deploying a Rust library on Android
+
+In our previous post we built a Rust library, and linked to it on [Android using JNI](https://mozilla.github.io/firefox-browser-architecture/experiments/2017-09-21-rust-on-android.html). This post builds on that by showing how to build the same app using JNA instead of JNI.
+
+# Why JNA?
+
+With JNI, for every `extern "C"` function that we created for our FFI, we had to create a mirror function in Rust for JNI to access. This involved some complicated function name construction and forced us to duplicate large chunks of code. JNA does away with all of that. You create a single interface that declares your FFI functions in Java and that's about it. It removes a large chunk of the headache and it is what we have moved to using here at Mozilla for our Android FFI.
+
+# Getting set up.
+
+Follow the instructions in [our previous post](https://mozilla.github.io/firefox-browser-architecture/experiments/2017-09-21-rust-on-android.html) for building and creating your own standalone toolchains and linking them to Rustup. Once you have your environment set up, head back here for what to do next.
```

Comment:
Later in this post there's a reference to a _greetings_ directory that is the parent of the _android_ directory, but this post doesn't describe creating it.

The previous post does describe doing so, as a precursor to creating the standalone toolchains. But folks that don't need to do that (either because they're on a platform where it isn't necessary or because they've already done it globally) might not follow the instructions in the previous blog post.

So I recommend including the instructions for creating the _greetings_ directory here as well, i.e.:

>First create a directory for our project.
>
>```
>mkdir greetings
>cd greetings
>```

Reply:
I think I might have missing this as I was moving stuff around in the post. Thanks for spotting.